lib/orderbook/OrderBook.ts

```diff
@@ -170,49 +192,114 @@ class OrderBook extends EventEmitter {
     }
   }
 
-  public addLimitOrder = (order: orders.OwnOrder): matchingEngine.MatchingResult => {
-    return this.addOwnOrder(order);
+  public addLimitOrder = async (order: orders.OwnOrder, ee?: EventEmitter): Promise<PlaceOrderResult> => {
+    const stampedOrder = this.stampOwnOrder(order);
+    return this.addOwnOrder(stampedOrder, false, ee, Date.now() + OrderBook.MAX_ADD_OWN_ORDER_ITERATIONS_TIME);
   }
 
-  public addMarketOrder = (order: orders.OwnMarketOrder): matchingEngine.MatchingResult => {
-    const price = order.isBuy ? Number.MAX_VALUE : 0;
-    const result = this.addOwnOrder({ ...order, price }, true);
+  public addMarketOrder = async (order: orders.OwnMarketOrder, ee?: EventEmitter): Promise<PlaceOrderResult> => {
+    const stampedOrder = this.stampOwnOrder({ ...order, price: order.isBuy ? Number.MAX_VALUE : 0 });
+    const result = await this.addOwnOrder(stampedOrder, true, ee, Date.now() + OrderBook.MAX_ADD_OWN_ORDER_ITERATIONS_TIME);
     delete result.remainingOrder;
     return result;
   }
 
-  private addOwnOrder = (order: orders.OwnOrder, discardRemaining = false): matchingEngine.MatchingResult => {
-    if (order.localId === '') {
-      // we were given a blank local id, so generate one
-      order.localId = uuidv1();
-    } else if (this.localIdMap.has(order.localId)) {
-      throw errors.DUPLICATE_ORDER(order.localId);
+  private addOwnOrder = async (order: orders.StampedOwnOrder, discardRemaining = false, ee?: EventEmitter, maxTime?: number):
+    Promise<PlaceOrderResult> => {
+    // this method can be called recursively on swap failures retries.
+    // if max time exceeded, don't try to match
+    if (maxTime && Date.now() > maxTime) {
+      assert(discardRemaining, 'discardRemaining must be true on recursive calls where maxTime could exceed');
```

Comment:
Why this restriction? Couldn't we just add the remaining order to the order book after time runs out?

Reply:
The idea is that the recursive calls don't add to the order book. They can add more `internalMatches` & `swapResults`, but the final `remainingOrder` is rolled back to the initial call, which will add it.